bn-apps/memberships-management/membership-service/src/main/kotlin/net/corda/businessnetworks/membership/member/NotifyMembersFlowResponder.kt

```diff
@@ -17,44 +14,30 @@ import net.corda.core.utilities.unwrap
  * Responder to the [NotifyMemberFlow]. The flow updates memberships cache with notifications from BNO
  */
 @InitiatedBy(NotifyMemberFlow::class)
-class NotifyMembersFlowResponder(private val session : FlowSession) : FlowLogic<Unit>(){
+class NotifyMembersFlowResponder(private val session : FlowSession) : FlowLogic<Unit>() {
 
     @Suspendable
     override fun call() {
-        val configuration = serviceHub.cordaService(MemberConfigurationService::class.java)
-        if (configuration.bnoParty() != session.counterparty) {
-            throw FlowException("Only BNO can start OnMembershipRevokedFlow")
-        }
+        val bno = session.counterparty
+
+        // don't forget to make sure that the messages are actually coming from the accepted BNOs
+        throwExceptionIfNotBNO(bno, serviceHub)
 
         val membershipService = serviceHub.cordaService(MembershipsCacheHolder::class.java)
         val notification = session.receive<Any>().unwrap { it }
         val cache = membershipService.cache
-        if (cache != null) {
-            when (notification) {
-                // removes suspended memberships from the cache
-                is OnMembershipSuspended -> {
-                    if (notification.suspendedMember != ourIdentity) {
-                        cache.suspendMembership(notification.suspendedMember)
-                    } else {
-                        membershipService.resetCache()
-                    }
-                }
-                // updates membership in the cache
-                is OnMembershipChanged -> {
-                    if (notification.changedMembership.state.data.member != ourIdentity) {
-                        cache.updateMembership(notification.changedMembership)
-                    }
+        when (notification) {
+            is OnMembershipChanged -> {
+                val membership = notification.changedMembership.state.data
+                // if our membership was suspended - then cleaning up the cache as suspended members won't get notifications anymore
+                if (membership.member == ourIdentity && membership.isSuspended()) {
+                    cache.reset(membership.bno)
+                } else {
+                    cache.updateMembership(notification.changedMembership)
```

Comment:
I guess it is going to work but it would be good to have a test showing that if I get revoked and later reactivated I do end up with the current set of members in my cache. As opposed to not being aware of what happened whilst I was revoked.

Reply:
Good point. Added a test.